### DISCUSSION TITLE
Array.prototype.slice() 에서 prototype이 무엇인가요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#41] [프론트 기술 질문 공간]
![image](https://user-images.githubusercontent.com/28939278/233921417-8964136d-4c46-4410-b1cc-a675cbe6dd0c.png)
`Array.prototype.slice()` 에서 `prototype`이 무엇인가요?
질문자를 JavaScript를 처음 접하는 개발자라고 가정하여 설명해주세요.

**Comment by michoball** [comment]
javascript를 처음 접하는 개발자한테는 어려울 수 있는데 괜찮은지 먼저 물어보고 이해할 의지가 보이면 설명을 해주겠습니다.


여기서 prototype은 Array 라는 객체가 참조하고 있는 Array의 원형이 되는 즉 Array의 부모객체라고 할 수 있습니다. 
여
그래서 여기서 Array.prototype.slice()를 호출하면 Array라는 객체가 참조하고 있는 Array.prototype인 Array의 부모 객체 안에 있는 slice라는 메서드를 호출하는 격이 되겠습니다. 

이렇게 하는 이유는 new 키워드를 사용하여 새로운 Array 객체를 생성했을 때 Array가 가지고 있는 메서드를 상속?받아 재사용할 수 있도록 하여 객체의 확장성있게 사용하기 위함입니다.

**Comment by WooGie911** [comment]
먼저 JavaScript를 처음 접하는 개발자에게 
Array.prototype.slice() 의 뜻을 설명한다면
Array 에 Array의 prototype이 가지고 있는 slice() 함수 
라고 설명하겠습니다.

따라서 Array.prototype.slice() 에서 prototype이 무엇인지 설명한다면
Array라는 객체가 참조하고있는( prototype ) 즉 배열의 원형인 객체라고 설명하겠습니다.

**Comment by skyhanull** [comment]
Array.prototype.slice() 에서 prototype은 부모객체 생성자가 가지고있는 기본적인 메소드들 이라고 할 수 있습니다
만약 slice 메소드를 사용할 때 이미 array생성자 함수로 만들어지게 되는 것이라 array안에 있는 함수들을 그대로 내려받아 사용할 수 있게됩니다
라고 설명할 것 같습니다.

**Comment by BaekYuri** [comment]
Array.prototype은 Array 객체의 부모 객체입니다. 또한 Array.prototype.slice()는 Array.prototype의 .slice() 라는 함수를 호출하는 것입니다.
실제로 Array 객체에 slice라는 함수가 없어도, prototype에 slice 함수가 있다면 이를 가져와 사용합니다.

**Comment by javascriptbiggosoo** [comment]
- 우리가 array를 만들 때 이용하던 array literal은 내부적으로는 이미 JavaScript에 정의된 `Array` constructor 함수가 실행되는 것입니다.
```JSX
const arr = [1, 2, 3]
```
```JSX
const arr = new Array(1, 2, 3)
```

- `new` operator가 포함된 함수 호출은 특정한 obejct를 반환하게 되어있는데, 예제에서 literal 형태로 생성된 `arr` 또한 그런 object인 것 입니다.
![image](https://user-images.githubusercontent.com/28939278/234760224-8cfe37fd-da0b-4525-bd5b-efd94a4b7252.png)

- `arr` 배열(이며 동시에 object)는 `Array` constructor 함수에 정의된 `prototype`의 프로퍼티와 메소드를 상속받는 것을 확인할 수 있습니다.
- `arr` 배열(이며 동시에 object)에 우리가 `slice` 메소드를 직접 정의하지 않지만 사용할 수 있는 것은 바로 상속받은 이 `prototype`에서 받아 사용하는 것입니다.

### 결론: array literal울 통한 생성은 `Array` 라는 하나의 함수가 작동하는 것이며, 이때 생성된 각 array는, `Array` 함수의 프로퍼티인 `prototype`을 상속 받습니다. 모든 array는 여기서 내장 프로퍼티, 메소드를 사용할 수 있는 것입니다.

> 하지만 평소 쓰시던대로 리터럴 형식으로 array를 생성하는 것이 권장됩니다. `Array` 함수의 동작이 직관에 맞지 않는 부분이 있습니다. https://zerodice0.tistory.com/202#:~:text=%EC%A6%89%2C%20new%20Array%20%28%29%20%EB%A5%BC%20%EC%82%AC%EC%9A%A9%ED%95%98%EC%97%AC%20%EB%B0%B0%EC%97%B4%EC%9D%84%20%EC%B4%88%EA%B8%B0%ED%99%94%ED%95%98%EC%A7%80,%EB%95%8C%EB%AC%B8%EC%9D%B8%20%EB%93%AF%20%ED%95%98%EB%8B%A4.%20%EC%A7%81%EA%B4%80%EC%84%B1%2C%20%EA%B7%B8%EB%A6%AC%EA%B3%A0%20%EC%9D%BC%EA%B4%80%EC%84%B1%EC%9D%98%20%EB%AC%B8%EC%A0%9C%EB%9D%BC%EA%B3%A0%20%ED%95%B4%EC%95%BC%ED%95%98%EB%82%98.
